feat: Pragma docs (#52)
kkrt-labs/kakarot-docs: docs/ecosystem/oracles/pragma.md

# Pragma Oracle

Pragma is the decentralized, transparent and composable oracle network, leveraging state-of-the-art zero-knowledge cryptography. We partner with the biggest market makers and the most liquid exchanges who sign and timestamp their own high quality, robust data and send it directly on-chain.

Our data feeds are live on 🥕 Kakarot, where they leverage the power of the ZK-EVM. Our code has been audited by leading security researchers at Nethermind and we have more audits coming up soon.

We are currently deployed on the following addresses:

> | Network | Cairo Address                                                     | PragmaCaller Solidity Interface                                                                                                  |

Reviewer: ```suggestion
| Network | Cairo Address                                                     | PragmaCaller Solidity Contract                                                                                                  |
```

| ------- | ----------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
| Sepolia | 0x3a99b4b9f711002f1976b3973f4b2031fe6056518615ff0f4e6dd829f972764 | [0x7491cA3699701a187C1a17308338Ad0bA258B082](https://sepolia.kakarotscan.org/address/0x7491cA3699701a187C1a17308338Ad0bA258B082) |
| Mainnet | **⏳ Soon**                                                        | **⏳ Soon**                                                                                                                       |

# Consuming our Data Feeds

You can get started with Pragma in just a few minutes. This guide will walk you through the process of consuming data from Pragma's oracle network, either from a Solidity or Cairo contract.

You can find the list of supported assets in our main documentation [here](https://docs.pragma.build/Resources/Cairo%201/data-feeds/supported-assets).

> ## Cairo

Reviewer: I would remove this from our docs as they're EVM-focused. wdyt @Eikix ?

Reviewer 2: Agreed, unless you have specific Kakarot data feeds in your docs. Maybe if you have a tag in your docs to indicate that a specific feed is supported on Kakarot or if we support 100% of your data feeds?

Author: Atm we use the same data feeds id in Kakarot so the list of supported assets work for the EVM too
I think @enitrat was talking about the `Cairo` code snippet?

Also, we're thinking about deploying one "feed" contract per pair to align with what Chainlink does for compatibility, I may update that soon

Reviewer: yes im talking about the cairo part.
@akhercha Considering that your architecture is entirely different would this not just add extra complexity?

Author: AFAIK it should be fine? We're just gonna build wrappers (one per pair/feed) around the `PragmaCaller` with the `ChainlinkAggregatorV3` interface

EDIT: something like that:
https://gist.github.com/akhercha/1c39a08f322d03950403095315bedfef

Reviewer: yep, i'm just wondering whether it makes sense to do this because in the end it all ends up being redirected to the same contract, the only thing you simplify is the ascii encoding of the pair

Reviewer 3: it's only to make it easy for chainlink consumers to migrate to pragma as a first step, even if it doesnt make that much sense it will make their life easier initially while porting to kkrt.

### 0. (Optional) Add Pragma as a dependency to your scarb/snforge project

```sh
scarb add pragma_lib --git https://github.com/astraly-labs/pragma-lib
```

### 1. Retrieve the BTC/USD Spot Median Price

```rust
use pragma_lib::abi::{IPragmaABIDispatcher, IPragmaABIDispatcherTrait};
use pragma_lib::types::{AggregationMode, DataType, PragmaPricesResponse};
use starknet::ContractAddress;
use starknet::contract_address::contract_address_const;

// felt252 conversion of "BTC/USD", can also write const KEY : felt252 = 'BTC/USD';
const KEY: felt252 = 18669995996566340;

fn get_asset_price_median(oracle_address: ContractAddress, asset: DataType) -> u128  {
    let oracle_dispatcher = IPragmaABIDispatcher{contract_address: oracle_address};
    let output: PragmaPricesResponse = oracle_dispatcher.get_data(asset, AggregationMode::Median(()));
    return output.price;
}

// USAGE
let oracle_address: ContractAddress = contract_address_const::<0x3a99b4b9f711002f1976b3973f4b2031fe6056518615ff0f4e6dd829f972764>();
let price = get_asset_price_median(oracle_address, DataType::SpotEntry(KEY));
```

## Solidity

> The Pragma Oracle cairo contract has been whitelisted by Kakarot - thus it is possible to call it directly in your Solidity contracts using the [PragmaCaller](https://github.com/kkrt-labs/kakarot/blob/main/solidity_contracts%2Fsrc%2FCairoPrecompiles%2FPragmaCaller.sol) interface!

Reviewer: ```suggestion
The Pragma Oracle cairo contract has been integrated by Kakarot - thus it is possible to call it directly in your Solidity contracts using the [PragmaCaller](https://github.com/kkrt-labs/kakarot/blob/main/solidity_contracts%2Fsrc%2FCairoPrecompiles%2FPragmaCaller.sol) contract!
```

### 1. Define the PragmaCaller interface

```typescript
interface IPragmaCaller {
    enum DataType {
        SpotEntry,
        FuturesEntry,
        GenericEntry
    }

    struct DataRequest {
        DataType dataType;
        uint256 pairId;
        uint256 expirationTimestamp;
    }

    struct PragmaPricesResponse {
        uint256 price;
        uint256 decimals;
        uint256 last_updated_timestamp;
        uint256 num_sources_aggregated;
        uint256 maybe_expiration_timestamp;
    }

    function getDataMedianSpot(DataRequest memory request) external view returns (PragmaPricesResponse memory);
}
```

### 2. Retrieve the BTC/USD Spot Median Price

```typescript
interface IPragmaCaller {
    // ... (include the interface definition from step 1)
}

contract CallerExample {
    IPragmaCaller private pragmaCaller;

>     // felt252 conversion of "BTC/USD";

Reviewer: ```suggestion
    // ascii conversion of "BTC/USD";
```

# Pragma Oracle

Pragma is the decentralized, transparent and composable oracle network, leveraging state-of-the-art zero-knowledge cryptography. We partner with the biggest market makers and the most liquid exchanges who sign and timestamp their own high quality, robust data and send it directly on-chain.

Our data feeds are live on 🥕 Kakarot, where they leverage the power of the ZK-EVM. Our code has been audited by leading security researchers at Nethermind and we have more audits coming up soon.

We are currently deployed on the following addresses:

> | Network | Cairo Address                                                     | PragmaCaller Solidity Contract                                                                                                   |

Reviewer 2: I would suggest to put first the PragmaCaller Solidity Contract, since most readers of the docs are Solidity devs

And Cairo Address isn't so clear for non-experts, I'd suggest calling it `Native Starknet Address (ignore if you're an EVM developer)`

Author: Should be good!

| ------- | ----------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- |

> | Sepolia | 0x3a99b4b9f711002f1976b3973f4b2031fe6056518615ff0f4e6dd829f972764 | [0x7491cA3699701a187C1a17308338Ad0bA258B082](https://sepolia.kakarotscan.org/address/0x7491cA3699701a187C1a17308338Ad0bA258B082) |

Reviewer 2: ```suggestion
| Kakarot Sepolia | 0x3a99b4b9f711002f1976b3973f4b2031fe6056518615ff0f4e6dd829f972764 | [0x7491cA3699701a187C1a17308338Ad0bA258B082](https://sepolia.kakarotscan.org/address/0x7491cA3699701a187C1a17308338Ad0bA258B082) |
```

> | Mainnet | **⏳ Soon**                                                        | **⏳ Soon**                                                                                                                       |

Reviewer 2: ```suggestion
| Kakarot Mainnet | **⏳ Soon**                                                        | **⏳ Soon**                                                                                                                       |
```

# Consuming our Data Feeds

You can get started with Pragma in just a few minutes. This guide will walk you through the process of consuming data from Pragma's oracle network.

## Solidity

To use our Price feeds, you have two choices:

## 1. Use our port of the `ChainlinkAggregatorV3` interface

If you already have an existing process working with Chainlink data feeds, you can use our `PragmaAggregatorV3` interfaces (under the hood, it calls the `PragmaCaller`).

The list of supported assets is:

> | Ticker     | Pair ID                  | Decimals | Sepolia                                                                                                                          | Mainnet    |

Reviewer 2: ```suggestion
| Ticker     | Pair ID                  | Decimals | Kakarot Sepolia                                                                                                                          | Mainnet    |
```

        uint256 num_sources_aggregated;
        uint256 maybe_expiration_timestamp;
    }

    function getDataMedianSpot(DataRequest memory request) external view returns (PragmaPricesResponse memory);
}
```

#### 2. Retrieve the BTC/USD Spot Median Price

```typescript
interface IPragmaCaller {
    // ... (include the interface definition from step 1)
}

contract CallerExample {
    IPragmaCaller private pragmaCaller;
    // ascii conversion of "BTC/USD";
    uint256 constant BTC_USD_FEED = 18669995996566340;

    constructor(address pragmaCallerAddress) {
        pragmaCaller = IPragmaCaller(pragmaCallerAddress);
    }

    function getDataMedianSpot(uint256 pairId) public view returns (IPragmaCaller.PragmaPricesResponse memory) {
        IPragmaCaller.DataRequest memory request = IPragmaCaller.DataRequest(
            IPragmaCaller.DataType.SpotEntry,
            pairId,
            0
        );
        return pragmaCaller.getDataMedianSpot(request);
    }

    function getBtcMedianPrice() public view returns (uint256) {
        IPragmaCaller.PragmaPricesResponse memory response = getDataMedianSpot(BTC_USD_FEED);
        return response.price;
    }
}
```

> An example has been deployed [here](https://sepolia.kakarotscan.org/address/0x9acb5dbE6B9E3569f4Ab8b4bf8E09F7efC330A26).

Reviewer 2: Can you verify the contract so that they can see the code directly on the explorer? We can help with that

Reviewer 2: Maybe adding a small github repository can go a long way too! People love to check out a "examples" repo

Author: > Can you verify the contract so that they can see the code directly on the explorer? We can help with that

All contracts are actually verified but since yesterday I can no longer access the "Contract" tab on the explorer - not sure why

Reviewer 2: Ah weird
